### PR TITLE
Update topology layouts to return a promise and to fire an event when done

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
/cc @andrewballantyne

### CODE REVIEW

File: frontend/packages/topology/src/layouts/ForceSimulation.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    // first remove the links so that the layout doesn't error
+    this.forceLink.links([]);
+
+    this.simulation.nodes([...nodes]);
+    this.forceLink.links([...links]);
+
+    return new Promise((resolve) => {
+      this.simulation.on('end', () => {
+        this.simulation.on('end', () => {});
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/layouts/ForceSimulation.ts]: Wouldn't this just add a second listener? Surely there can be multiple listeners on a given event...

File: frontend/packages/topology/src/layouts/DagreLayout.ts [hunk 2]
```diff
@@ -1,119 +1,142 @@
...
+      this.element.setBendpoints(
+        this.points.slice(1, -1).map((point: any) => new Point(point.x, point.y)),
+      );
+    }
   }
+}
 
-  get y(): number {
-    return this.node.getBounds().getCenter().y;
+type DagreLayoutOptions = LayoutOptions & {
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/layouts/DagreLayout.ts]:
```suggestion
type DagreLayoutOptions = LayoutOptions & dagre.GraphLabel;
```
`dagre.GraphLabel` is all the options that the graph can take.

File: frontend/packages/topology/src/layouts/DagreLayout.ts [hunk 3]
```diff
@@ -1,119 +1,142 @@
...
+      ...{
+        marginx: 0,
+        marginy: 0,
+        nodesep: this.options.nodeDistance,
+        edgesep: this.options.linkDistance,
+        ranker: 'tight-tree',
+        rankdir: 'TB',
+      },
+      ...options,
+    };
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/layouts/DagreLayout.ts]:
```
    this.dagreOptions = {
      ...this.options,
      marginx: 0,
      marginy: 0,
      nodesep: this.options.nodeDistance,
      edgesep: this.options.linkDistance,
      ranker: 'tight-tree',
      rankdir: 'TB',
      ...options,
    };
```

File: frontend/packages/topology/src/layouts/DagreLayout.ts [hunk 4]
```diff
@@ -1,119 +1,142 @@
...
+    if (initialRun || addingNodes) {
+      const dagreGraph = new dagre.graphlib.Graph({ compound: true });
+      dagreGraph.setGraph({
+        marginx: this.dagreOptions.marginx,
+        marginy: this.dagreOptions.marginy,
+        nodesep: this.dagreOptions.nodesep,
+        edgesep: this.dagreOptions.edgesep,
+        ranker: this.dagreOptions.ranker,
+        rankdir: this.dagreOptions.rankdir,
+      });
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/layouts/DagreLayout.ts]: I think we need a way to apply all options I send you as long as it's valid Dagre options. I'm using `align` and `ranksep` for instance, to do my layout.

**Comment by christianvogt** [comment]
Why do we need layout to return a promise?

**Comment by jeff-phillips-18** [comment]
At some point, layouts could contain nodes with layouts and the parent will depend on the layout completing before it can start its layout. This is a precursor to that and enables any other caller to handle the completion without necessarily using events.

**Comment by jeff-phillips-18** [comment]
Closing... no need to return promises at this time.

**Pull request closed without merging**